### PR TITLE
imx: cortexa7: support TechNexion PICO-PI-IMX7

### PR DESCRIPTION
**Pull request by Leo-PL**
TechNexion PICO-PI-IMX7 is a Freescale i.MX7D-based development board in
the well-known "Raspberry Pi" form factor, comprising of PICO-IMX7 SoM
and the PICO-PI-IMX7 carrier board.
Usually bundled with 5" 800x480 LVDS display with I2C touch screen,
and an Omnivision OV5645 camera on a MIPI CSI bus, on a daughterboard.
The board was previously used primarily with "Android Things" ecosystem,
but the project was killed by Google.
    
This would not be possible, if not for the great tutorial of
setting up Debian on this board, by Robert C. Nelson [1].

This series also includes a minor refactoring of the imx target, required to support other i.MX7-based boards in future.

Hardware highlights:
CPU: Freescale i.MX7D SoC, dual-core Cortex-A7 at 1000MHz
RAM: 512MiB DDR3 SDRAM
Storage: 4GB eMMC
Networking:
- Built-in Gigabit Ethernet with Atheros AR8035 PHY,
- Broadcom BCM4339 1x1 802.11ac Wi-Fi (over SDIO) + Bluetooth 4.1 (over
  SDIO+UART+IS2) combo, with Hirose u.FL connector on the board,
- Dual CAN interfaces on the 40-pin connector,
Interfaces:
- USB-C power input plus USB 2.0 OTG host/device port,
- Single USB-A host port,
- Serial console over built-in FT232BL USB-UART converter with micro-USB
  connector. Configuration: 115200-8-N-1,
- Analog audio interface with TRRS connector in CTIA standard,
- SPI, I2C and UART interfaces available on the 40-pin and,
  the mikroBUS connector,
- I2C connector for the optional touch panel,
- Parallel LCD output for the optional display,
- MIPI CSI connector for the optional camera.

Installation:
0. Connect the serial console to debug USB connector and the terminal of
   choice in another window, at 115200-8-N-1.
   Ensure you can switch to it quickly after next step.
1. Power-on the board from your PC. Ensure your PC can supply the
   required current, the board can take more than 1A in the peak load
   during booting, and brownout will result in power-on reset loop.
   Preferably, use charging-capable USB port, or connect through
   self-powered USB hub.
   If U-boot is present already on the system, interrupt boot sequence
   by pressing any key and skip to point 6.
2. Ensure the boot mode jumpers J1 and J2 are in correct position for USB recovery:
```
       2   6  2   6
      --------------
      |o o-o||o-o o|
      |o o-o||o-o o|
   J1 -------------- J2
       1   5  1   5
```
   The jumpers are located just underneath the 40-pin expansion header,
   and are of the smaller 2mm pitch.
3. Download and build imx_usb_loader from
   https://github.com/boundarydevices/imx_usb_loader.
3. Power-on the board again from your PC through USB OTG connector
4. Use imx_usb_loader to load SPL and u-boot-dtb.img to the board:
```
   $ sudo imx_usb u-boot-pico-pi-imx7d/SPL
   $ sudo imx_usb u-boot-pico-pi-imx7d/u-boot-dtb.img
```
5. Switch to the terminal from step 2, and interrupt boot sequence by
   pressing any key within 2 seconds.
6. Configure mmc 0 to boot from the data partition,
   and disable access to boot partitions:
```
   => mmc partconf 0 0 7 0
```
   This only needs to be set once, if you were running Debian
   previously, this is probably already set.
7. Enable USB mass storage passthrough for eMMC from U-boot
```
   => ums 0 mmc 0
```
8. Optionally, backup previous eMMC contents by reading out its image
9. Copy over the factory image to the USB device, for example:
```
   $ sudo dd if=openwrt-imx-cortexa7-pico-pi-imx7d-squashfs.combined.bin \
     of=/dev/disk/by-id/usb-Linux_UMS_disk_0-0:0 \
     bs=8M status=progress oflag=direct
```
10. Detach USB MSC interface from your PC an U-boot by pressing Ctrl+C
1`. Ensure, that boot mode jumpers are at the default settings for eMMC boot:
```
       2   6  2   6
      --------------
      |o-o o||o o-o|
      |o-o o||o-o o|
   J1 -------------- J2
       1   5  1   5
```
   If they are not, power-off the board, restore them,
   and power-on the board again.
   Otherwise, if jumpers are set, just reset the board from U-boot:
```
   => reset
```

The installation is now complete, and board should boot successfully.

Upgrading: Just use sysupgrade image, as usual in OpenWrt.

Known issues/current limitations:
- OV5645 camera - not described in upstream device tree as of kernel
  5.15. As of v5.18.y there are staging drivers present in upstream
  Linux tree for i.MX CSI interface, together with new bindings, so this
  expected to get supported eventually.
- on-chip ADCs are disabled in upstream device tree, so the the kernel
  driver remains disabled as well.

[1] https://forum.digikey.com/t/debian-getting-started-with-the-pico-pi-imx7/12429

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>

### CODE REVIEW

**Comment by Leo-PL** [comment]
cc @pepe2k and @ynezz

**Comment by pepe2k** [comment]
@Leo-PL @ynezz I have this board, I will take care of this one.

Cheers,
Piotr

File: package/boot/uboot-imx/Makefile [hunk 1]
```diff
@@ -42,10 +42,19 @@ define U-Boot/wandboard
   BUILD_DEVICES:=wandboard_dual
 endef
 
+define U-Boot/pico-pi-imx7d
```
Comment (pepe2k) [inline package/boot/uboot-imx/Makefile:45]: @Leo-PL please, manage alphabetical order of devices definitions.

File: package/boot/uboot-envtools/files/imx_cortexa7 [hunk 2]
```diff
@@ -0,0 +1,23 @@
+#
+# Copyright (C) 2013-2014 OpenWrt.org
```
Comment (pepe2k) [inline package/boot/uboot-envtools/files/imx_cortexa7:2]: @Leo-PL 2022 :)

File: target/linux/imx/cortexa7/config-default [hunk 3]
```diff
@@ -1,27 +1,35 @@
...
 CONFIG_MTD_SPI_NOR=y
 CONFIG_MTD_SPLIT_FIRMWARE=y
 CONFIG_MTD_SPLIT_FIT_FW=y
-CONFIG_PINCTRL_IMX=y
-CONFIG_PINCTRL_IMX6UL=y
+CONFIG_NOP_USB_XCEIV=y
 CONFIG_PM_GENERIC_DOMAINS=y
 CONFIG_PM_GENERIC_DOMAINS_OF=y
-CONFIG_SOC_IMX6=y
-CONFIG_SOC_IMX6UL=y
```
Comment (pepe2k) [inline target/linux/imx/cortexa7/config-default:24]: @Leo-PL The i.MX6UL/L series is based on the Cortex-A7, I'm planning to include several boards within this subtarget. Any reason why did you remove these configs?

File: target/linux/imx/cortexa7/config-default [hunk 4]
```diff
@@ -11,6 +11,8 @@ CONFIG_CMA_AREAS=7
 # CONFIG_CMA_SYSFS is not set
 CONFIG_CONTIG_ALLOC=y
 # CONFIG_DMA_CMA is not set
+CONFIG_EXTRA_FIRMWARE="imx/sdma/sdma-imx7d.bin"
```
Comment (pepe2k) [inline target/linux/imx/cortexa7/config-default:34]: @Leo-PL I think we need to switch to sysfs based solution for this subtarget as at least the i.MX6UL/L will require `sdma-imx6q`. I will handle that when adding other devices, let's keep it as it for now.

File: package/kernel/linux/modules/other.mk [hunk 5]
```diff
@@ -64,6 +64,24 @@ endef
 
 $(eval $(call KernelPackage,bluetooth))
 
+define KernelPackage/btsdio
+  SUBMENU:=$(OTHER_MENU)
+  TITLE:=Bluetooth support
```
Comment (pepe2k) [inline package/kernel/linux/modules/other.mk:69]: We probably should mention SDIO in title.

File: target/linux/imx/cortexa7/config-default (hunk 4 above)
Comment (Leo-PL) [reply target/linux/imx/cortexa7/config-default:34]: Not only that, this opens a potential can of worms of linking closed-source SDMA firmware in GPL-licensed kernel.

File: target/linux/imx/image/common.mk [hunk 6]
```diff
@@ -49,3 +49,13 @@ define Build/imx-sdcard
 
 	$(Build/imx-combined-image-clean)
 endef
+
+define Build/imx-sdcard-raw-uboot
```
Comment (pepe2k) [inline target/linux/imx/image/common.mk:53]: @Leo-PL isn't the default configuration for SPL for this board capable of starting U-Boot from e.g. ext filesystem? This way only the SPL is burn in a raw mode on eMMC and rest is easily accessible in `boot` partition.

File: target/linux/imx/cortexa7/config-default (hunk 3 above)
Comment (Leo-PL) [reply target/linux/imx/cortexa7/config-default:24]: I wasn't aware of that only until recently - of course this should stay, at least CONFIG_SOC_IMX6UL for sure.

File: package/kernel/linux/modules/other.mk (hunk 5 above)
Comment (Leo-PL) [reply package/kernel/linux/modules/other.mk:69]: Correct, I'll fix that soon.

File: target/linux/imx/cortexa7/base-files/lib/upgrade/platform.sh [hunk 7]
```diff
@@ -1,16 +1,59 @@
-PART_NAME=firmware
-REQUIRE_IMAGE_METADATA=1
+#
+# Copyright (C) 2010-2015 OpenWrt.org
```
Comment (pepe2k) [inline target/linux/imx/cortexa7/base-files/lib/upgrade/platform.sh:2]: 2022 :)

File: target/linux/imx/image/common.mk (hunk 6 above)
Comment (Leo-PL) [reply target/linux/imx/image/common.mk:53]: I tried that and it didn't work, but I may have written incorrect artifact there. I presume that I just need to put u-boot.bin on /boot and that's it, like it is done already for Apalis eMMC?

**Comment by Leo-PL** [comment]
Thanks for review, I'll fix the issues when I return home, or tomorrow ;-)

File: target/linux/imx/image/cortexa7.mk [hunk 8]
```diff
@@ -10,3 +12,21 @@ define Device/Default
   KERNEL_LOADADDR := 0x80008000
   IMAGES :=
 endef
+
+define Device/pico-pi-imx7d
```
Comment (pepe2k) [inline target/linux/imx/image/cortexa7.mk:16]: @Leo-PL please make it `vendor_product`. This way you could also drop `SUPPORTED_DEVICES`.

File: target/linux/imx/image/cortexa7.mk [hunk 9]
```diff
@@ -10,3 +12,21 @@ define Device/Default
...
+  SUPPORTED_DEVICES := technexion,imx7d-pico-pi
+  UBOOT := pico-pi-imx7d
+  DEVICE_DTS := imx7d-pico-pi
+  DEVICE_PACKAGES := kmod-sound-core kmod-sound-soc-imx kmod-sound-soc-imx-sgtl5000 \
+	kmod-can kmod-can-flexcan kmod-can-raw kmod-leds-gpio \
+	kmod-input-touchscreen-edt-ft5x06 kmod-usb-hid kmod-btsdio \
+	kmod-brcmfmac cypress-firmware-4339-sdio cypress-nvram-4339-pico-pi-imx7d
+  FILESYSTEMS := squashfs
+  IMAGES := combined.bin sysupgrade.bin
+  DEVICE_IMG_NAME = $$(DEVICE_IMG_PREFIX)-$$(1).$$(2)
```
Comment (pepe2k) [inline target/linux/imx/image/cortexa7.mk:28]: @Leo-PL any reason why you didn't use generic image filename?

File: target/linux/imx/image/common.mk (hunk 6 above)
Comment (pepe2k) [reply target/linux/imx/image/common.mk:53]: Not sure about Apalis but during my tests of UDOO Neo board (SD, not eMMC), SPL was capable of booting `uboot-dtb.img` from the `/boot` partition and this recipe worked just fine: `IMAGE/combined.bin := append-rootfs | pad-extra 128k | imx6-sdcard`

Comment (pepe2k) [reply target/linux/imx/image/common.mk:53]: Have a look at a0528cab445ad6bb4b372d4484bac10f638c433f

File: target/linux/imx/cortexa7/config-default (hunk 4 above)
Comment (pepe2k) [reply target/linux/imx/cortexa7/config-default:34]: To be honest I never looked at license side of these firmware files.

Comment (Leo-PL) [reply target/linux/imx/cortexa7/config-default:34]: Recently at my day job, my team took a look at this, and this was the conclusion - we switched to FIT images to easily bundle it in initramfs instead.

Comment (pepe2k) [reply target/linux/imx/cortexa7/config-default:34]: @Leo-PL what do you mean by `bundle` here? Include it as a `loadable` binary blob within FIT?

File: target/linux/imx/image/cortexa7.mk (hunk 9 above)
Comment (Leo-PL) [reply target/linux/imx/image/cortexa7.mk:28]: Just copy-paste from Apalis, IIRC. This can probably be dropped.

File: target/linux/imx/cortexa7/config-default (hunk 3 above)
Comment (Leo-PL) [reply target/linux/imx/cortexa7/config-default:24]: I restored all i.MX6UL-related configs back - will post update soon, after I rebuild and test.

File: target/linux/imx/image/common.mk (hunk 6 above)
Comment (Leo-PL) [reply target/linux/imx/image/common.mk:53]: Tried again and didn't work, despite I've seen both `u-boot-dtb.img` and `u-boot.img` in include/configs/imx7_spl.h in CONFIG_SPL_FS_LOAD_PAYLOAD_NAME - because neither CONFIG_SPL_FS_FAT nor CONFIG_SPL_FS_EXT4 is set for that board in upstream U-boot. So I guess this image should stay.

**Comment by Leo-PL** [comment]
@pepe2k I posted updated version with fixes for the comments, ready for another round of review.

File: target/linux/imx/image/cortexa7.mk (hunk 8 above)
Comment (Leo-PL) [reply target/linux/imx/image/cortexa7.mk]: Makes total sense - done.

File: target/linux/imx/cortexa7/base-files/lib/upgrade/platform.sh (hunk 7 above)
Comment (Leo-PL) [reply target/linux/imx/cortexa7/base-files/lib/upgrade/platform.sh]: Fixed.

File: package/boot/uboot-envtools/files/imx_cortexa7 (hunk 2 above)
Comment (Leo-PL) [reply package/boot/uboot-envtools/files/imx_cortexa7]: Fixed.

File: package/boot/uboot-imx/Makefile [hunk 10]
```diff
@@ -42,10 +42,19 @@ define U-Boot/wandboard
...
+  BUILD_SUBTARGET:=cortexa7
+  BUILD_DEVICES:=pico-pi-imx7d
+endef
+
 UBOOT_TARGETS := \
 	apalis_imx6 \
 	mx6cuboxi \
-	wandboard
+	wandboard \
+	pico-pi-imx7d
```
[the remark this answers is not in the document]
Comment (Leo-PL) [reply package/boot/uboot-imx/Makefile]: Fixed.

File: package/boot/uboot-imx/Makefile (hunk 1 above)
Comment (Leo-PL) [reply package/boot/uboot-imx/Makefile:39]: Fixed.

**Comment by Leo-PL** [comment]
Today I found out, that the drivers for MIPI CSI and OV5645 camera are actually available in 5.15 kernel, so I attempted to make it run. However, during that I spotted that device tree isn't updated by sysupgrade, presumably on all i.MX targets, and I believe it should be updateable as well, together with the kernel.
@pepe2k what do you think about extending sysupgrade image generation from generic `sysupgrade-tar`? Again, it is one thing that using FIT images for booting would solve trivially.

**Comment by pepe2k** [comment]
@Leo-PL

> I posted updated version with fixes for the comments, ready for another round of review.

Thanks, everything looks good, sorry for late reply. Just one more nitpick - I think it's safe to force whole subtarget (this will be first device anyway) to include metadata in image by default, thus no reason to make `REQUIRE_IMAGE_METADATA` variable per device.

> Today I found out, that the drivers for MIPI CSI and OV5645 camera are actually available in 5.15 kernel, so I attempted to make it run. However, during that I spotted that device tree isn't updated by sysupgrade, presumably on all i.MX targets, and I believe it should be updatable as well, together with the kernel.

Yes, I have the same opinion here.

> @pepe2k what do you think about extending sysupgrade image generation from generic `sysupgrade-tar`? Again, it is one thing that using FIT images for booting would solve trivially.

Personally, I would prefer to just use FIT based kernel image. I did some work in that terms for UDOO Neo (I will rebase and share it tomorrow in my staging tree).

But in case of your device, you would need to add `CONFIG_FIT=y` to the U-Boot defconfig with a local patch. And as we are going to do that, I would also enable support for loading U-Boot from filesystem by SPL we discussed before and... send that trivial patch to upstream afterwards :grin: What do you think?

> In the meantime, I'll update the commit message so information about CSI is correct. I don't have even the roughest estimate on camera support, so let's not block this PR waiting on it.

Agree.

Cheers,
Piotr

**Comment by Leo-PL** [comment]
@pepe2k makes total sense, let's do that.
While at that, let's enable CONFIG_REGULATOR_ANATOP as well, which was causing the MIPI CSI core not to probe for me, maybe I'll get that camera to work.

Stay tuned :-)

**Comment by Leo-PL** [comment]
@pepe2k please ping me (if possible) when you post your work on FIT images, I'll pick them up. For now, I'll continue doing the cleanup regarding drivers to enable, I got some more previously missing stuff to probe, but still not the dreaded `imx_thermal` thing.

**Comment by pepe2k** [comment]
@Leo-PL it seems I need more time for that FIT related code rework as it turned out what I'm planning to achieve requires changes across whole tree and some run-time tests in other targets.

I don't want to block this one, so I will just start merging your changes and later we can improve/fix.

> Any idea which kernel symbol I might be missing, besides `CONFIG_ARM_IMX_CPUFREQ_DT` and obvious `CONFIG_IMX_THERMAL` to enable CPU frequency scaling?

I will look into that and let you know.

**Comment by pepe2k** [comment]
@Leo-PL if you haven't yet squashed all your local changes, please push them as separated commits on top of those in PR. I have already merged everything locally and added some other fixes.

**Comment by Leo-PL** [comment]
@pepe2k I have them squashed, so I'll have to rebase them anyway - so I think it is best if I rebase them once PR lands into master. But don't worry, they were relatively minor.

**Comment by pepe2k** [comment]
@Leo-PL side question, what do you use for kernel config change/update? Are you doing that manually, directly in `config-default`? During merging your changes I had to perform config refresh after every your commit which touches subtarget kernel configuraiton (with `make kernel_{defconfig,menuconfig} CONFIG_TARGET=...`) because of missing some hidden symbols selected automatically and also some added which have unmet dependencies and would get removed/deselected anyway.

**Comment by Leo-PL** [comment]
Yes, I did it by hand - because `make kernel_menuconfig` touched config for whole target, not only for `cortexa7`.

**Comment by Leo-PL** [comment]
@pepe2k are there any changes I should introduce before merging, maybe rebase against current master?

**Comment by pepe2k** [comment]
@Leo-PL sorry for the delay, I did rebasing and also some other minor fixes and changes (e.g. commits reorder, consistency with device name, proper kernel config changes).

For now I keep everything in my staging tree (I still have few small things on TODO list for this one) and plan to push it upstream during next days: https://git.openwrt.org/?p=openwrt/staging/pepe2k.git;a=shortlog;h=refs/heads/pr-9802

If you have time you can double check and let me know if everything is fine.

Cheers,
Piotr

**Comment by Leo-PL** [comment]
@pepe2k LGTM.
